enhancements/proposals/1978-scylladbdatacenter-api/README.md

```diff
@@ -0,0 +1,724 @@
+  resources:
+    requests:
+      cpu: 1
+      memory: 4Gi
+      storage: 100Gi
+scyllaDBManagerAgent:
+  image: docker.io/scylladb/scylla-manager-agent:3.3.0
+  resources:
+    requests:
+      cpu: 100m
+      memory: 100Mi
+nodes: 9
+racks:
+- name: a
+- name: b
+- name: c
+
+```
+
+### API Spec
+
+Proposed API spec:
+```
+
+const (
+    AvailableCondition   = "Available"
+    ProgressingCondition = "Progressing"
+    DegradedCondition    = "Degraded"
+)
+
+// ScyllaDBDatacenterSpec defines the desired state of ScyllaDBDatacenter.
+type ScyllaDBDatacenterSpec struct {
+    // metadata controls shared metadata for all pods created based on this spec.
+    // +optional
+    Metadata *ObjectTemplateMetadata `json:"metadata,omitempty"`
+
+    // datacenterName is the name of the scylla datacenter. Used as datacenter name in GossipingPropertyFileSnitch.
+    DatacenterName string `json:"datacenterName"`
+
+    // scyllaDB holds a specification of ScyllaDB.
+    ScyllaDB ScyllaDB `json:"scyllaDB"`
+
+    // scyllaManagerAgent holds a specification of ScyllaDB Manager Agent.
+    // +optional
+    ScyllaDBManagerAgent *ScyllaDBManagerAgent `json:"scyllaDBManagerAgent,omitempty"`
+
+    // imagePullSecrets is an optional list of references to secrets in the same namespace
+    // used for pulling any images used by this spec.
+    // +optional
+    ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
+
+    // network holds network configuration.
+    // +optional
+    Network *Network `json:"network,omitempty"`
+
+    // dnsDomains specifies a list of DNS domains this cluster is reachable by.
+    // These domains are used when setting up the infrastructure, like certificates.
+    // +optional
+    DNSDomains []string `json:"dnsDomains,omitempty"`
+
+    // forceRedeploymentReason can be used to force a rolling restart of all racks in this DC by providing a unique string.
+    // +optional
+    ForceRedeploymentReason string `json:"forceRedeploymentReason,omitempty"`
+
+    // exposeOptions specifies parameters related to exposing ScyllaCluster backends.
+    // +optional
+    ExposeOptions *ExposeOptions `json:"exposeOptions,omitempty"`
+
+    // nodes specify desired total number of nodes in this datacenter.
+    // This number will be balanced out across all specified racks.
+    // Cannot be used together with `nodes` specified within any rack spec.
+    // +optional
+    Nodes *int32 `json:"nodes"`
+
+    // racks specify the racks in the datacenter.
+    // +optional
+    Racks []RackSpec `json:"racks"`
+
+    // automaticOrphanedNodeCleanup controls if automatic orphan node cleanup should be performed.
+    AutomaticOrphanedNodeCleanup bool `json:"automaticOrphanedNodeCleanup,omitempty"`
+
+    // sysctls holds the sysctl properties to be applied during initialization given as a list of key=value pairs.
+    // Example: fs.aio-max-nr=232323
+    // +optional
+    Sysctls []string `json:"sysctls,omitempty"`
```

Comment:
racks/zones can have different machine types - maybe this should be on the rack level - i guess a similar point can be made for some other fields on this level

Reply:
I haven't look deep enough to see whether some sysctls make sense per container, but I see that the majority should be per node. So maybe this would go away with https://github.com/scylladb/scylla-operator/issues/749. I agree that internal annotation let's us defer the decision. (Btw. it would be good to add all those reasons to the annotation const definitions, so we know when we can drop it.)

Reply:
I suppose we should file issues to track deprecation of fields that you are moving to annotation, unless they already have one like hostNetworking.

Reply:
Once we have everything finalized I will create issues for missing parts